### DISCUSSION TITLE
Pseudo-Measurements

### DISCUSSION THREAD

**Discussion by scud-soptim** [PowerGridModel/.github#28] [Q&A] [answered]
In state estimation, a pseudo-balance measurement (e.g., S=0 MVar) is often used at a node without a load or generator to achieve more stable results. Unfortunately, I couldn't find specific references to this in the source code of PGM (presumably Power Grid Management).

1. Does PGM have a similar implementation for pseudo-balance measurements?
2. Would adding additional balance measurements improve the performance of state estimation in PGM?

Thank you for your insights and assistance!

**Comment by TonyXiang8787** [comment, marked as answer]
Hi @sudo-ac thanks for your question. Indeed we treat a node without any load/generator as a zero-injection constraint. We enforce this constraints in our solver. The user does not need to add any sensors to achieve this. The PGM can automatically detect a zero-injection node and enforce the constraint.

If you are interested, please see below the code segment.

https://github.com/PowerGridModel/power-grid-model/blob/cdb65fce5b24a6ff1f6ccfd8c39e14bf45e8fb04/power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/measured_values.hpp#L234-L235

**Reply by scud-soptim** [reply]
Thank you for the answer. Is there a possibility to give Zero-Measurement a small variation?

**Reply by TonyXiang8787** [reply]
What do you mean exactly "a small variation"?

**Reply by mgovers** [reply]
If I understand the problem correctly, you may want to add a [Shunt](https://power-grid-model.readthedocs.io/en/stable/user_manual/components.html#shunt) component to provide any additional loss of power.

**Reply by scud-soptim** [reply]
No, what I meant is to introduce a pseudo-measurement of 0 MVar at a bus without loads or generators. This pseudo-measurement should not be treated as a constraint but rather as a high-fidelity measurement, significantly enhancing the accuracy of the state estimation. This approach is often instrumental in stabilizing the state estimation process.

**Reply by mgovers** [reply]
right, I see what you mean. Apologies for misunderstanding.

We do not currently support this particular use case in the way you mention it now. This is for two reasons:

* by design, we assume the equations regarding electrical engineering to be exact.
  * Any deviation implies hidden features of the grid, which should be modeled as a load/generator or a shunt
* it makes computations easier (faster)
  * it greatly reduces the dimensionality of the state estimation solver

I think the closest thing to what you intend to do is to:

* add a load or generator to the node in question (this addresses reason 1)
* and then use a power sensor connected to that appliance with measured value `0 M Var` and `power_sigma` set to whichever accuracy you prefer. (this addresses reason 2)

Just for the sake of completeness, this is also related to the topics of

* observability (see also https://github.com/PowerGridModel/power-grid-model/issues/84 and https://github.com/PowerGridModel/power-grid-model/issues/508 )
* partial injection constraints (e.g. power sensor on links which we currently disallow by design; see also https://github.com/PowerGridModel/power-grid-model/issues/164 and https://github.com/PowerGridModel/power-grid-model/issues/35 ).

**Reply by TonyXiang8787** [reply]
Hi @sudo-ac, I do not fully understand the reasoning of such a pseudo-measurement you proposed. Usually the equation system is more stable if we incorporate the zero-injection as constraints. You state that a pseudo-measurement with high-fidelity can enhance the accuracy and stability. It appears to me that usually the opposite is true. Can you give some concrte numerical examples?

**Reply by scud-soptim** [reply]
I obtained this information from the book 'Netzberechnung' by Prof. K. Schäfer, Chapter Pseudomessungen, pages 322ff. Similarly, from the article by G. Beißler, 'Behandlung passiver Knoten bei der Zustandserkennung,' etzArchive, Vol. 3, No. 6, pp. 179–184, 1981. We also employ a state estimation technique that applies this method. However, passive nodes can fundamentally be treated as constraints. Unfortunately, I don't have any specific numerical examples.

**Comment by scud-soptim** [comment]
thank you very much 👍
